### PR TITLE
hal: Support various range types as function parameters

### LINKED ISSUE #1749: Equivalent to VK_WHOLE_SIZE
Vulkan allows passing `VK_WHOLE_SIZE` (basically `-1`) as value. We could use an enum for handling this case.

### PR DESCRIPTION
**Pull request by msiglreith**
Adds the Range abstraction discussed in #1749 
Supporting `..`, `a..`, `..b`, `a..b` and `(Option<_>, Option<_>)` (covering all prior mentioned cases).

**WIP**: Update gl, d3d12, metal and higher level crates

Fixes #1749

### CODE REVIEW

**Comment by omni-viral** [comment]
Consider to use [Bound](https://doc.rust-lang.org/std/collections/enum.Bound.html) type insted of `Option`.

**Comment by msiglreith** [comment]
I used it in the first iteration but went with `Option` as it results in less code in the backend implementation. Inclusive ranges are not part of stable rust yet, which leaves us with two instead of three bound options atm.

**Comment by omni-viral** [comment]
@msiglreith `Bound` itself is stable. So the user may specify inclusive range with it.
Using `Option` may cause a slight confusion because (Some(x), Some(y))` is the same as `x..y` meaning from `x` included to `y` excluded.
But it may be not an issue at all.

File: src/backend/gl/src/device.rs [hunk 1]
```diff
@@ -702,25 +706,27 @@ impl d::Device<B> for Device {
         }
     }
 
-    fn flush_mapped_memory_ranges<'a, I>(&self, _: I)
+    fn flush_mapped_memory_ranges<'a, I, R>(&self, _: I)
```
[the remark this answers is not in the document]
Comment (msiglreith) [reply src/backend/gl/src/device.rs]: One can also fallback to `(Option<u64>, Option<u64>` if needed

**Comment by msiglreith** [comment]
Finished the other backends beside metal. Trying to change it without `cargo check` and travis OS X build times takes too long. Would someone mind updating the backend?

**Comment by kvark** [comment]
@msiglreith I'll see what I can do ;)

**Comment by kvark** [comment]
Bors r+

**Pull request merged**